rancher/rancher
# [v2.9] Fix system agent upgrader logic for Windows nodes

## Issue: https://github.com/rancher/rancher/issues/46396 (2.9 Backport, https://github.com/rancher/rancher/issues/46451)

## Problem

Rancher v2.9.0 introduced the Agent TLS Mode feature, but this feature was not fully implemented on Windows nodes. Unfortunately, some of the logic required to implement this feature for Windows nodes has never worked properly. Before this PR, it has not been possible to deploy System Agent Upgrader Plans to Windows nodes, as the image used in those plans did not have the ability to modify the host. The `rancher/wins` image is being updated alongside this PR, and will be included in Rancher when a new tag is cut. In order for the Agent TLS Mode feature to be supported on Windows nodes, the way that Rancher deploys System Agent Upgrade plans on Windows nodes needs to be revised.
 
## Solution
In order to resolve this quite a few things have to change

1. The `rancher/wins` image, intended to be used by the system agent upgrader, is being refactored to run in a Host Process pod and the install script must detect the STRICT_VERIFY variable. This is being done in https://github.com/rancher/wins/pull/225
2. Properly define the STRICT_VERIFY environment variable when crafting the windows install script
3. Update the System Upgrade Controller API dependency to allow for the configuration of a Security Context. The Security Context must then be configured to support Host Process Pods.
4. Ensure that SUC plans are only created if the latest version of the System Upgrade Controller has been deployed onto the downstream cluster. This ensures that plans are not processed by a version of the SUC which does not support all of the configured fields (as is the case on 2.9.1). 
5. Only create SUC plans for Windows nodes which are running at least the August patches of rke2. Prior versions of rke2 on windows contain a bug which can break pod networking if the rke2 service is restarted. Currently, each time that rancher-wins is restarted the rke2 service is restarted as well.


## Testing

+ Deploy a 2.8 version of Rancher using Helm. I used 2.8.5 
+ Create an rke2 Windows cluster running the august patches (1.29.8)
+ Ensure the cluster becomes available
+ Update the rancher deployment to a version containing these changes
+ Ensure that the SUC plan is properly deployed to windows nodes once the SUC has been fully upgraded, and that the plan and resulting pods have the correct security context
+ Ensure that the rancher-wins SUC image properly configures the host, (i.e. updates the rancher-wins config file (`c:/etc/rancher/wins/config`))
+ Ensure that the cluster is operational
+ Update an agent environment variable (`CATTLE_WINS_DEBUG`). Ensure that an additional Windows SUC plan is deployed and completes successfully. Ensure that the rancher-wins config file is updated (`c:/etc/rancher/wins/config`)

Do the same as above but using a version of rke2 older than the August patches. Ensure that SUC plans are not deployed until the cluster is upgraded to at least the August patches

## Engineering Testing
### Manual Testing

I have done the above tests using development builds of Rancher and rancher-wins. 

### Automated Testing

* Test types added/modified:
    * Unit
       *  A new unit test has been added to ensure that we limit plan deployments to specific Windows rke2 versions
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

## QA Testing Considerations

This should be tested heavily. The test plan for the agent tls mode feature must be redone with a Windows cluster. We should also test cluster upgrades from versions which are not expected to have SUC plans deployed onto them (all versions prior to the August patches) to versions which are expected to have SUC plans (August patches and beyond). Scenarios where Rancher is not configured with CA certs, but later enables them during an upgrade, should also be tested. 

### Regressions Considerations
Ensure that the Linux SUC plans continue to behave as expected and work under all circumstances, particularly right after an upgrade of the Rancher version. 

Existing / newly added automated tests that provide evidence there are no regressions:
